lib/rails_state_machine/state_machine.rb

```diff
@@ -59,7 +64,11 @@ def event(name, &block)
     end
 
     def model_constant(name, value)
-      @model.const_set(name, value)
+      if @model.const_defined?(name.to_s)
+        raise(StateAlreadyDefinedError, "State :#{value} has already been defined in another state_machine. Please use \"prefix\" as a keyword argument.")
+      else
+        @model.const_set(name, value)
+      end
```

Comment:
Earlier, I wrote:

> The prefix must be respected in all relevant cases.
> In doing so, you'll notice that a duplication check needs to happen elsewhere.

I still believe that here is not the correct place to check for state duplication.

The logic should live in a separate method that is called in `configure` or similar. We already have `has_state?`; you can use that one the check if any of the states that would be added already exist.

Reply:
Good advice, thanks! But I don't think that `has_state?` can be used for that, since I need to check if another state machine has already defined the same state. As far as I know, I can only check that by checking if there's already a constant with that name. (I could probably also check, if the helper methods have already been defined, but I prefer the `constant_defined?` approach.) 

I would define a method like that:
```ruby
def check_if_states_already_defined
  @states_by_name.each do |state_name, _|
    if @model.const_defined?("STATE_#{state_name.upcase}")
      raise(StateAlreadyDefinedError, "State :#{state_name} has already been defined in another state_machine. Please use \"prefix\" as a keyword argument.")
    end
  end
end
```

and add it to the beginning of `configure`, right after the `instance_eval`. Is that okay?

Reply:
That's too much internal API for my taste, and still bound to the constants while they are only one aspect of name duplication.

We can already fetch all other state machines of a model via `model.state_machines`, and each state machine offers `has_state?`. I'm confident something like this :point_down: should work:

```ruby
other_state_machines = model.state_machines.except(state_attribute).values
other_state_macihnes.any? { |state_machine| state_machine.has_state?(state_name) }
```